server/shared/src/main/scala/org/http4s/server/middleware/Throttle.scala

```diff
@@ -53,40 +56,50 @@ object Throttle {
+      Temporal[F].monotonic.map(_.toNanos)
+
     /** Creates an in-memory [[TokenBucket]].
       *
-      * @param capacity the number of tokens the bucket can hold and starts with.
+      * @param capacity    the number of tokens the bucket can hold and starts with.
       * @param refillEvery the frequency with which to add another token if there is capacity spare.
       * @return A task to create the [[TokenBucket]].
       */
     def local[F[_]](capacity: Int, refillEvery: FiniteDuration)(implicit
         F: Temporal[F]
-    ): F[TokenBucket[F]] = {
-      def getTime = F.monotonic.map(_.toNanos)
-      val bucket = getTime.flatMap(time => F.ref((capacity.toDouble, time)))
-
-      bucket.map { counter =>
-        new TokenBucket[F] {
-          override def takeToken: F[TokenAvailability] =
-            counter.access.flatMap { case ((previousTokens, previousTime), setter) =>
-              getTime.flatMap { currentTime =>
-                val timeDifference = currentTime - previousTime
-                val tokensToAdd = timeDifference.toDouble / refillEvery.toNanos.toDouble
-                val newTokenTotal = Math.min(previousTokens + tokensToAdd, capacity.toDouble)
-                def unavailable = {
-                  val timeToNextToken = refillEvery.toNanos - timeDifference
-                  TokenUnavailable(timeToNextToken.nanos.some)
-                }
-
-                // If setter fails (yields the value false), then retry with recursive call
-                if (newTokenTotal >= 1)
-                  setter((newTokenTotal - 1, currentTime)).ifM(F.pure(TokenAvailable), takeToken)
-                else
-                  setter((newTokenTotal, currentTime)).ifM(F.pure(unavailable), takeToken)
+    ): F[TokenBucket[F]] = for {
+      // toNanos throws if called on an infinite duration
+      nanoValue <- F.catchNonFatal(refillEvery.toNanos)
+      // make sure that refillEvery is positive
+      _ <- F.raiseUnless(nanoValue > 0L)(
+        new IllegalArgumentException("refillEvery should be > 0 nano")
+      )
+      creationTime <- getTime
+      counter <- F.ref((capacity.toDouble, creationTime))
+    } yield new TokenBucket[F] {
+      override def takeToken: F[TokenAvailability] =
+        for {
+          values <- counter.access
+          ((previousTokens, previousTime), setter) = values
+          currentTime <- getTime
+          token <- {
+            val timeDifference = currentTime - previousTime
+            val tokensToAdd = timeDifference.toDouble / refillEvery.toNanos.toDouble
+            val newTokenTotal = Math.min(previousTokens + tokensToAdd, capacity.toDouble)
+
+            // If setter fails (yields the value false), then retry with recursive call
+            if (newTokenTotal >= 1)
+              setter((newTokenTotal - 1, currentTime))
+                .ifM(F.pure(TokenAvailable: TokenAvailability), takeToken)
+            else {
+              lazy val unavailable: TokenAvailability = {
```

Comment:
Nit but this was fine as a `def`.`lazy val` can have hidden performance costs due to synchronization.

Reply:
I tend to prefer `lazy val` to `def` whenever the evaluation of the expression can happen _at most once_. Also, in this case I don't see the potential synchronization issue, as it's a local (helper) variable that could have well been inline'd at call site, hence my proposal to change it as is. I don't mind leaving it as a `def` anyway 😊 will fix

Reply:
> I tend to prefer lazy val to def whenever the evaluation of the expression can happen at most once. 

Actually, when I know this for sure, I would definitely prefer a `def` because there would be no benefit to caching a `lazy val`, but only potential performance issues due to the overhead.

Just to illustrate that, here's how the code decompiles with `lazy val`

```java
    private final /* synthetic */ Throttle.TokenAvailability unavailable$lzycompute$1(LazyRef unavailable$lzy$1, long timeDifference$1) {
        Throttle.TokenAvailability tokenAvailability;
        LazyRef lazyRef = unavailable$lzy$1;
        synchronized (lazyRef) {
            Throttle.TokenAvailability tokenAvailability2;
            if (unavailable$lzy$1.initialized()) {
                tokenAvailability2 = (Throttle.TokenAvailability)unavailable$lzy$1.value();
            } else {
                long timeToNextToken = this.refillEvery$1.toNanos() - timeDifference$1;
                tokenAvailability2 = (Throttle.TokenAvailability)unavailable$lzy$1.initialize((Object)new Throttle.TokenUnavailable((Option<FiniteDuration>)OptionIdOps$.MODULE$.some$extension(implicits$.MODULE$.catsSyntaxOptionId((Object)new package.DurationLong(package$.MODULE$.DurationLong(timeToNextToken)).nanos()))));
            }
            tokenAvailability = tokenAvailability2;
        }
        return tokenAvailability;
    }
```

Here's the code if we make it a `def` instead.

```java
    private final Throttle.TokenAvailability unavailable$1(long timeDifference$1) {
        long timeToNextToken = this.refillEvery$1.toNanos() - timeDifference$1;
        return new Throttle.TokenUnavailable((Option<FiniteDuration>)OptionIdOps$.MODULE$.some$extension(implicits$.MODULE$.catsSyntaxOptionId((Object)new package.DurationLong(package$.MODULE$.DurationLong(timeToNextToken)).nanos())));
    }
```

For sure I'm no expert, but I think the second one has a better chance at being JITed and inlined by the JVM.

Reply:
I'm no expert either, and I hear what you're saying. I've reverted it to a def :+1: